File: v11/interpreters/ts/scripts/test-cli.ts
import fs from "node:fs";
import { promises as fsPromises } from "node:fs";
import path from "node:path";
import { fileURLToPath } from "node:url";

import { AST, TypeChecker, V11 } from "../index";
import type { TypecheckerDiagnostic } from "../src/typechecker/diagnostics";
import { callCallableValue } from "../src/interpreter/functions";
import { makeIntegerValue } from "../src/interpreter/numeric";
import { ExitSignal } from "../src/interpreter/signals";
import { ParserDiagnosticError } from "../src/parser/diagnostics";

import { ModuleLoader, type Program } from "./module-loader";
import { collectModuleSearchPaths, type ModuleSearchPath } from "./module-search-paths";
import { buildExecutionSearchPaths, loadManifestContext } from "./module-deps";
import { discoverRoot, indexSourceFiles } from "./module-utils";
import { normalizeRepoRelativePath } from "./path-utils";
import { buildRuntimeDiagnostic, formatParserDiagnostic, formatRuntimeDiagnostic, formatTypecheckerDiagnostic, printPackageSummaries } from "./typecheck-utils";
import type { TypecheckMode } from "./typecheck-mode";

export type TestReporterFormat = "doc" | "progress" | "tap" | "json";

export type TestCliFilters = {
  includePaths: string[];
  excludePaths: string[];
  includeNames: string[];
  excludeNames: string[];
  includeTags: string[];
  excludeTags: string[];
};

export type TestRunOptions = {
  failFast: boolean;
  repeat: number;
  parallelism: number;
  shuffleSeed?: number;
};

export type TestCliConfig = {
  targets: string[];
  filters: TestCliFilters;
  run: TestRunOptions;
  reporterFormat: TestReporterFormat;
  listOnly: boolean;
  dryRun: boolean;
};

type HarnessFailure = {
  message: string;
  details: string | null;
};

type RootTestPlan = {
  rootDir: string;
  rootName: string;
  entryFile: string;
  testFiles: string[];
  packageNames: Set<string>;
};

export type TestLoadResult = {
  programs: Program[];
  modules: Program["modules"];
};

type ModuleDiagnosticEntry = {
  packageName: string;
  diagnostic: TypecheckerDiagnostic;
};

type TestCliEnv = {
  ablePathEnv: string;
  ableModulePathsEnv: string;
};

const TEST_FILE_SUFFIXES = [".test.able", ".spec.able"];

export async function resolveTestTargets(targets: string[]): Promise<string[]> {
  const rawTargets = targets.length > 0 ? targets : ["."];
  const resolved: string[] = [];
  const seen = new Set<string>();

  for (const target of rawTargets) {
    const abs = path.resolve(process.cwd(), target);
    let stats: fs.Stats;
    try {
      stats = await fsPromises.stat(abs);
    } catch (error) {
      throw new Error(`unable to access ${abs}: ${(error as Error).message}`);
    }
    if (stats.isFile()) {
      if (isTestFile(abs)) {
        if (!seen.has(abs)) {
          seen.add(abs);
          resolved.push(abs);
        }
      } else {
        const dir = path.dirname(abs);
        if (!seen.has(dir)) {
          seen.add(dir);
          resolved.push(dir);
        }
      }
      continue;
    }
    if (!stats.isDirectory()) {
      throw new Error(`unsupported test target: ${abs}`);
    }
    if (!seen.has(abs)) {
      seen.add(abs);
      resolved.push(abs);
    }
  }

  return resolved;
}

export async function collectTestFiles(targets: string[]): Promise<string[]> {
  const found = new Set<string>();
  for (const target of targets) {
    const stats = await fsPromises.stat(target);
    if (stats.isFile()) {
      if (isTestFile(target)) {
        found.add(path.resolve(target));
      }
      continue;
    }
    if (stats.isDirectory()) {
      await walkTestFiles(target, found);
    }
  }
  return [...found].sort();
}

export async function loadTestPrograms(
  testFiles: string[],
  env: TestCliEnv,
): Promise<TestLoadResult | null> {
  let plans: RootTestPlan[];
  try {
    plans = await buildRootTestPlans(testFiles);
  } catch (error) {
    console.error(`able test: ${extractErrorMessage(error)}`);
    return null;
  }

  if (plans.length === 0) {
    return { programs: [], modules: [] };
  }

  const searchPaths = await resolveTestSearchPaths(plans, env);
  if (!searchPaths) {
    return null;
  }

  const programs: Program[] = [];
  for (const plan of plans) {
    const includePackages = new Set<string>([
      "able.test.harness",
      "able.test.protocol",
      "able.test.reporters",
      ...plan.packageNames,
    ]);
    const loader = new ModuleLoader(searchPaths);
    try {
      const program = await loader.load(plan.entryFile, { includePackages: [...includePackages] });
      programs.push(program);
    } catch (error) {
      if (error instanceof ParserDiagnosticError) {
        console.error(formatParserDiagnostic(error.diagnostic, { absolutePath: true }));
        return null;
      }
      console.error(`failed to load tests from ${plan.entryFile}: ${extractErrorMessage(error)}`);
      return null;
    }
  }

  return {
    programs,
    modules: mergeTestModules(programs),
  };
}

export function maybeTypecheckTestModules(
  session: TypeChecker.TypecheckerSession,
  modules: Program["modules"],
  typecheckMode: TypecheckMode,
): boolean {
  if (typecheckMode === "off") {
    return true;
  }
  const diagnostics: ModuleDiagnosticEntry[] = [];
  const seen = new Set<string>();
  for (const mod of modules) {
    if (seen.has(mod.packageName)) {
      continue;
    }
    seen.add(mod.packageName);
    const result = session.checkModule(mod.module);
    if (result.diagnostics.length === 0) {
      continue;
    }
    const packageName =
      result.summary?.name ?? mod.packageName ?? resolveModulePackageName(mod.module);
    if (isStdlibPackage(packageName)) {
      continue;
    }
    const filtered = result.diagnostics.filter((diag) => !isStdlibDiagnostic(diag));
    if (filtered.length === 0) {
      continue;
    }
    for (const diag of filtered) {
      diagnostics.push({ packageName, diagnostic: diag });
    }
  }
  if (diagnostics.length === 0) {
    return true;
  }
  emitDiagnostics(diagnostics);
  printPackageSummaries(session.getPackageSummaries());

  if (typecheckMode === "strict") {
    process.exitCode = 2;
    return false;
  }

  console.warn("typechecker: proceeding despite diagnostics because ABLE_TYPECHECK_FIXTURES=warn");
  return true;
}

export async function evaluateTestModules(interpreter: V11.Interpreter, modules: Program["modules"]): Promise<boolean> {
  const evaluated = new Set<string>();
  for (const mod of modules) {
    if (evaluated.has(mod.packageName)) {
      continue;
    }
    evaluated.add(mod.packageName);
    try {
      await interpreter.evaluateAsTask(mod.module);
    } catch (error) {
      if (error instanceof ExitSignal) {
        process.exitCode = error.code;
        return false;
      }
      console.error(formatRuntimeFailure(error));
      process.exitCode = 2;
      return false;
    }
  }
  return true;
}

export function buildDiscoveryRequest(interpreter: V11.Interpreter, config: TestCliConfig): V11.RuntimeValue {
  const def = getStructDef(interpreter, "able.test.protocol", "DiscoveryRequest");
  return interpreter.makeNamedStructInstance(def, [
    ["include_paths", makeStringArray(interpreter, config.filters.includePaths)],
    ["exclude_paths", makeStringArray(interpreter, config.filters.excludePaths)],
    ["include_names", makeStringArray(interpreter, config.filters.includeNames)],
    ["exclude_names", makeStringArray(interpreter, config.filters.excludeNames)],
    ["include_tags", makeStringArray(interpreter, config.filters.includeTags)],
    ["exclude_tags", makeStringArray(interpreter, config.filters.excludeTags)],
    ["list_only", { kind: "bool", value: config.listOnly }],
  ]);
}

export function buildRunOptions(interpreter: V11.Interpreter, run: TestRunOptions): V11.RuntimeValue {
  const def = getStructDef(interpreter, "able.test.protocol", "RunOptions");
  const shuffleSeed =
    run.shuffleSeed === undefined ? { kind: "nil", value: null } : makeIntegerValue("i64", BigInt(run.shuffleSeed));
  return interpreter.makeNamedStructInstance(def, [
    ["shuffle_seed", shuffleSeed],
    ["fail_fast", { kind: "bool", value: run.failFast }],
    ["parallelism", makeIntegerValue("i32", BigInt(run.parallelism))],
    ["repeat", makeIntegerValue("i32", BigInt(run.repeat))],
  ]);
}

export function buildTestPlan(interpreter: V11.Interpreter, descriptors: V11.RuntimeValue): V11.RuntimeValue {
  const def = getStructDef(interpreter, "able.test.protocol", "TestPlan");
  return interpreter.makeNamedStructInstance(def, [["descriptors", descriptors]]);
}

export function callHarnessDiscover(
  interpreter: V11.Interpreter,
  request: V11.RuntimeValue,
): V11.RuntimeValue | null {
  const discover = getCallableSymbol(interpreter, "able.test.harness", "discover_all");
  if (!discover) {
    console.error("able test: unable to find able.test.harness.discover_all");
    return null;
  }
  try {
    const result = callCallableValue(interpreter as any, discover, [request], interpreter.globals);
    const failure = extractFailure(interpreter, result);
    if (failure) {
      console.error(`able test: ${formatFailure(failure)}`);
      return null;
    }
    if (result.kind !== "array" && !isKernelArray(result)) {
      console.error("able test: discovery returned unexpected result");
      return null;
    }
    return result;
  } catch (error) {
    console.error(`able test: ${extractErrorMessage(error)}`);
    return null;
  }
}

export async function callHarnessRun(
  interpreter: V11.Interpreter,
  plan: V11.RuntimeValue,
  options: V11.RuntimeValue,
  reporter: V11.RuntimeValue,
): Promise<HarnessFailure | null> {
  const runPlan = getCallableSymbol(interpreter, "able.test.harness", "run_plan");
  if (!runPlan) {
    console.error("able test: unable to find able.test.harness.run_plan");
    return { message: "missing able.test.harness.run_plan", details: null };
  }
  const callEnv = new V11.Environment(interpreter.globals);
  const runPlanIdent = "__able_test_run_plan";
  const planIdent = "__able_test_plan";
  const optionsIdent = "__able_test_options";
  const reporterIdent = "__able_test_reporter";
  callEnv.define(runPlanIdent, runPlan);
  callEnv.define(planIdent, plan);
  callEnv.define(optionsIdent, options);
  callEnv.define(reporterIdent, reporter);
  const callNode = AST.functionCall(AST.identifier(runPlanIdent), [
    AST.identifier(planIdent),
    AST.identifier(optionsIdent),
    AST.identifier(reporterIdent),
  ]);
  try {
    const result = await interpreter.evaluateAsTask(callNode, callEnv);
    if (result.kind === "nil") {
      return null;
    }
    const failure = extractFailure(interpreter, result);
    if (failure) {
      console.error(`able test: ${formatFailure(failure)}`);
      return failure;
    }
    console.error("able test: run_plan returned unexpected result");
    return { message: "run_plan returned unexpected result", details: null };
  } catch (error) {
    console.error(`able test: ${extractErrorMessage(error)}`);
    return { message: extractErrorMessage(error), details: null };
  }
}

function isTestFile(pathname: string): boolean {
  return TEST_FILE_SUFFIXES.some((suffix) => pathname.endsWith(suffix));
}

async function walkTestFiles(dir: string, found: Set<string>): Promise<void> {
  const entries = await fsPromises.readdir(dir, { withFileTypes: true });
  for (const entry of entries) {
    const fullPath = path.join(dir, entry.name);
    if (entry.isDirectory()) {
      if (entry.name === "quarantine" || entry.name === "node_modules" || entry.name === ".git") {
        continue;
      }
      await walkTestFiles(fullPath, found);
      continue;
    }
    if (entry.isFile() && isTestFile(entry.name)) {
      found.add(path.resolve(fullPath));
    }
  }
}

async function buildRootTestPlans(testFiles: string[]): Promise<RootTestPlan[]> {
  const roots = new Map<string, RootTestPlan>();
  for (const file of testFiles) {
    const { rootDir, rootName } = await discoverRoot(file);
    const existing = roots.get(rootDir);
    if (existing) {
      if (existing.rootName !== rootName) {
        throw new Error(`conflicting package roots at ${rootDir}`);
      }
      existing.testFiles.push(file);
      continue;
    }
    roots.set(rootDir, {
      rootDir,
      rootName,
      entryFile: file,
      testFiles: [file],
      packageNames: new Set(),
    });
  }

  for (const plan of roots.values()) {
    const { fileToPackage } = await indexSourceFiles(plan.rootDir, plan.rootName);
    for (const file of plan.testFiles) {
      const pkgName = fileToPackage.get(file);
      if (!pkgName) {
        throw new Error(`unable to resolve package for test file ${file}`);
      }
      plan.packageNames.add(pkgName as string);
    }
  }

  return [...roots.values()];
}

async function resolveTestSearchPaths(
  plans: RootTestPlan[],
  env: TestCliEnv,
): Promise<ModuleSearchPath[] | null> {
  const extras: ModuleSearchPath[] = [];
  const probeFrom = new Set<string>();

  for (const plan of plans) {
    probeFrom.add(plan.rootDir);
    try {
      const manifestContext = await loadManifestContext(plan.rootDir);
      if (manifestContext.manifest?.path) {
        probeFrom.add(path.dirname(manifestContext.manifest.path));
      }
      extras.push(...buildExecutionSearchPaths(manifestContext.manifest, manifestContext.lock));
    } catch (error) {
      console.error(extractErrorMessage(error));
      return null;
    }
  }

  probeFrom.add(process.cwd());
  probeFrom.add(path.dirname(fileURLToPath(import.meta.url)));
  probeFrom.add(path.dirname(process.execPath));

  return collectModuleSearchPaths({
    cwd: process.cwd(),
    ablePathEnv: env.ablePathEnv,
    ableModulePathsEnv: env.ableModulePathsEnv,
    extras,
    probeFrom: [...probeFrom],
  });
}

function mergeTestModules(programs: Program[]): Program["modules"] {
  const modules: Program["modules"] = [];
  const seen = new Set<string>();
  for (const program of programs) {
    for (const mod of program.modules) {
      if (seen.has(mod.packageName)) {
        continue;
      }
      seen.add(mod.packageName);
      modules.push(mod);
    }
  }
  return modules;
}

function structField(value: Extract<V11.RuntimeValue, { kind: "struct_instance" }>, field: string): V11.RuntimeValue {
  if (value.values instanceof Map) {
    return value.values.get(field) ?? { kind: "nil", value: null };
  }
  const fieldIndex = value.def.fields.findIndex((entry) => entry.name?.name === field);
  if (fieldIndex >= 0 && Array.isArray(value.values)) {
    return value.values[fieldIndex] ?? { kind: "nil", value: null };
  }
  return { kind: "nil", value: null };
}

function coerceToString(interpreter: V11.Interpreter, value: V11.RuntimeValue | undefined): string {
  if (!value) return "";
  if (value.kind === "String") return value.value;
  try {
    return interpreter.valueToString(value);
  } catch {
    return String(value);
  }
}

function extractFailure(interpreter: V11.Interpreter, value: V11.RuntimeValue): HarnessFailure | null {
  if (!value || value.kind !== "struct_instance") {
    return null;
  }
  if (value.def.id.name !== "Failure") {
    return null;
  }
  const message = coerceToString(interpreter, structField(value, "message"));
  const detailsValue = structField(value, "details");
  const details = detailsValue.kind === "nil" ? null : coerceToString(interpreter, detailsValue);
  return { message, details };
}

function isKernelArray(value: V11.RuntimeValue): value is Extract<V11.RuntimeValue, { kind: "struct_instance" }> {
  return value.kind === "struct_instance" && value.def.id.name === "Array";
}

function formatFailure(failure: HarnessFailure): string {
  const detail = failure.details ? ` (${failure.details})` : "";
  return `${failure.message}${detail}`;
}

function getStructDef(
  interpreter: V11.Interpreter,
  packageName: string,
  name: string,
): AST.StructDefinition {
  const pkg = interpreter.packageRegistry.get(packageName);
  if (!pkg) {
    throw new Error(`unable to locate package ${packageName}`);
  }
  const value = pkg.get(name);
  if (!value || value.kind !== "struct_def") {
    throw new Error(`unable to locate ${name} in ${packageName}`);
  }
  return value.def;
}

function getCallableSymbol(
  interpreter: V11.Interpreter,
  packageName: string,
  name: string,
): V11.RuntimeValue | null {
  const pkg = interpreter.packageRegistry.get(packageName);
  if (!pkg) return null;
  return getCallableFromPackage(pkg, name);
}

function getCallableFromPackage(
  pkg: Map<string, V11.RuntimeValue>,
  name: string,
): V11.RuntimeValue | null {
  const value = pkg.get(name);
  if (!value) return null;
  if (
    value.kind === "function" ||
    value.kind === "function_overload" ||
    value.kind === "native_function" ||
    value.kind === "bound_method" ||
    value.kind === "native_bound_method" ||
    value.kind === "partial_function"
  ) {
    return value;
  }
  return null;
}

function makeStringArray(interpreter: V11.Interpreter, values: string[]): V11.RuntimeValue {
  return interpreter.makeArrayValue(
    values.map((value) => ({ kind: "String", value })),
  );
}

function resolveModulePackageName(module: AST.Module | undefined | null): string {
  const segments =
    module?.package?.namePath
      ?.map((identifier) => identifier?.name)
      .filter((segment): segment is string => Boolean(segment)) ?? [];
  if (segments.length === 0) {
    return "<anonymous>";
  }
  return segments.join(".");
}

function isStdlibPackage(packageName: string | undefined | null): boolean {
  if (!packageName) return false;
  return packageName === "able" || packageName.startsWith("able.");
}

function isStdlibDiagnostic(diag: TypecheckerDiagnostic): boolean {
  const rawPath = diag.location?.path;
  if (!rawPath) return false;
  const relative = normalizeRepoRelativePath(rawPath);
  if (relative.startsWith("..")) return false;
  return relative.startsWith("v11/stdlib/") || relative.startsWith("v11/kernel/");
}

function emitDiagnostics(diags: ModuleDiagnosticEntry[]): void {
  const seen = new Set<string>();
  for (const entry of diags) {
    const formatted = formatTypecheckerDiagnostic(entry.diagnostic, {
      packageName: entry.packageName,
      absolutePath: true,
    });
    if (seen.has(formatted)) {
      continue;
    }
    seen.add(formatted);
    console.error(formatted);
  }
}

function formatRuntimeFailure(error: unknown): string {
  return formatRuntimeDiagnostic(buildRuntimeDiagnostic(error), { absolutePath: true });
}

function extractErrorMessage(err: unknown): string {
  if (!err) return "";
  if (typeof err === "string") return err;
  if (err instanceof Error) {
    if (process.env.ABLE_TRACE_ERRORS) {
      const anyErr = err as any;
      const stack = err.stack ?? err.message;
      if (anyErr.value && typeof anyErr.value === "object" && "message" in anyErr.value) {
        return `${stack}\nRaised: ${String(anyErr.value.message)}`;
      }
      return stack;
    }
    const anyErr = err as any;
    if (anyErr.value && typeof anyErr.value === "object" && "message" in anyErr.value) {
      return String(anyErr.value.message);
    }
    return err.message;
  }
  if (typeof err === "object" && err) {
    const anyErr = err as any;
    if (typeof anyErr.message === "string") return anyErr.message;
  }
  return String(err);
}

File: v12/interpreters/go/cmd/able/test_cli.go
package main

import (
	"fmt"
	"io/fs"
	"os"
	"path/filepath"
	"sort"
	"strconv"
	"strings"
	"time"

	"able/interpreter-go/pkg/interpreter"
)

func runTest(args []string) int {
	config, err := parseTestArguments(args)
	if err != nil {
		fmt.Fprintf(os.Stderr, "able test: %v\n", err)
		return 1
	}

	targets, err := resolveTestTargets(config.Targets)
	if err != nil {
		fmt.Fprintf(os.Stderr, "able test: %v\n", err)
		return 1
	}

	testFiles, err := collectTestFiles(targets)
	if err != nil {
		fmt.Fprintf(os.Stderr, "able test: %v\n", err)
		return 1
	}

	if len(testFiles) == 0 {
		fmt.Fprintln(os.Stdout, "able test: no test modules found")
		return 0
	}

	loadResult, err := loadTestPrograms(testFiles)
	if err != nil {
		fmt.Fprintf(os.Stderr, "able test: %v\n", err)
		return 2
	}

	mode := resolveTestTypecheckMode()
	if ok, code := typecheckTestModules(loadResult.modules, mode); !ok {
		return code
	}

	interp := interpreter.New()
	registerPrint(interp)

	if ok, code := evaluateTestModules(interp, loadResult.modules); !ok {
		return code
	}

	cliModule, err := loadTestCliModule(interp)
	if err != nil {
		fmt.Fprintf(os.Stderr, "able test: %v\n", err)
		return 2
	}

	discoveryRequest, err := buildDiscoveryRequest(interp, cliModule, config)
	if err != nil {
		fmt.Fprintf(os.Stderr, "able test: %v\n", err)
		return 2
	}

	descriptors, ok := callHarnessDiscover(interp, cliModule, discoveryRequest)
	if !ok {
		return 2
	}

	if config.ListOnly || config.DryRun {
		emitTestPlanList(interp, descriptors, config)
		return 0
	}

	if arrayLength(interp, descriptors) == 0 {
		fmt.Fprintln(os.Stdout, "able test: no tests to run")
		return 0
	}

	state := &TestEventState{}
	reporter, err := createTestReporter(interp, cliModule, config.ReporterFormat, state)
	if err != nil {
		fmt.Fprintf(os.Stderr, "able test: %v\n", err)
		return 2
	}

	runOptions, err := buildRunOptions(interp, cliModule, config.Run)
	if err != nil {
		fmt.Fprintf(os.Stderr, "able test: %v\n", err)
		return 2
	}
	testPlan, err := buildTestPlan(cliModule, descriptors)
	if err != nil {
		fmt.Fprintf(os.Stderr, "able test: %v\n", err)
		return 2
	}

	if callHarnessRun(interp, cliModule, testPlan, runOptions, reporter.reporter) != nil {
		return 2
	}

	if reporter.finish != nil {
		reporter.finish()
	}

	if state.FrameworkErrors > 0 {
		return 2
	}
	if state.Failed > 0 {
		return 1
	}
	return 0
}

func parseTestArguments(args []string) (TestCliConfig, error) {
	filters := TestCliFilters{
		IncludePaths: []string{},
		ExcludePaths: []string{},
		IncludeNames: []string{},
		ExcludeNames: []string{},
		IncludeTags:  []string{},
		ExcludeTags:  []string{},
	}
	run := TestRunOptions{
		FailFast:    false,
		Repeat:      1,
		Parallelism: 1,
	}
	format := reporterDoc
	listOnly := false
	dryRun := false
	var shuffleSeed *int64
	var targets []string

	for i := 0; i < len(args); i++ {
		arg := args[i]
		switch arg {
		case "--list":
			listOnly = true
		case "--dry-run":
			dryRun = true
			listOnly = true
		case "--path":
			val, err := expectFlagValue(arg, nextArg(args, &i))
			if err != nil {
				return TestCliConfig{}, err
			}
			filters.IncludePaths = append(filters.IncludePaths, val)
		case "--exclude-path":
			val, err := expectFlagValue(arg, nextArg(args, &i))
			if err != nil {
				return TestCliConfig{}, err
			}
			filters.ExcludePaths = append(filters.ExcludePaths, val)
		case "--name":
			val, err := expectFlagValue(arg, nextArg(args, &i))
			if err != nil {
				return TestCliConfig{}, err
			}
			filters.IncludeNames = append(filters.IncludeNames, val)
		case "--exclude-name":
			val, err := expectFlagValue(arg, nextArg(args, &i))
			if err != nil {
				return TestCliConfig{}, err
			}
			filters.ExcludeNames = append(filters.ExcludeNames, val)
		case "--tag":
			val, err := expectFlagValue(arg, nextArg(args, &i))
			if err != nil {
				return TestCliConfig{}, err
			}
			filters.IncludeTags = append(filters.IncludeTags, val)
		case "--exclude-tag":
			val, err := expectFlagValue(arg, nextArg(args, &i))
			if err != nil {
				return TestCliConfig{}, err
			}
			filters.ExcludeTags = append(filters.ExcludeTags, val)
		case "--format":
			val, err := expectFlagValue(arg, nextArg(args, &i))
			if err != nil {
				return TestCliConfig{}, err
			}
			parsed, err := parseReporterFormat(val)
			if err != nil {
				return TestCliConfig{}, err
			}
			format = parsed
		case "--fail-fast":
			run.FailFast = true
		case "--repeat":
			val, err := expectFlagValue(arg, nextArg(args, &i))
			if err != nil {
				return TestCliConfig{}, err
			}
			count, err := parsePositiveInt(val, arg, 1)
			if err != nil {
				return TestCliConfig{}, err
			}
			run.Repeat = count
		case "--parallel":
			val, err := expectFlagValue(arg, nextArg(args, &i))
			if err != nil {
				return TestCliConfig{}, err
			}
			count, err := parsePositiveInt(val, arg, 1)
			if err != nil {
				return TestCliConfig{}, err
			}
			run.Parallelism = count
		case "--shuffle":
			next := peekArg(args, i+1)
			if next != "" && !strings.HasPrefix(next, "-") {
				seed, err := parsePositiveInt(next, arg, 0)
				if err != nil {
					return TestCliConfig{}, err
				}
				seedVal := int64(seed)
				shuffleSeed = &seedVal
				i++
			} else {
				seed := generateShuffleSeed()
				shuffleSeed = &seed
			}
		default:
			if strings.HasPrefix(arg, "-") {
				return TestCliConfig{}, fmt.Errorf("unknown able test flag '%s'", arg)
			}
			targets = append(targets, arg)
		}
	}

	run.ShuffleSeed = shuffleSeed

	return TestCliConfig{
		Targets:        targets,
		Filters:        filters,
		Run:            run,
		ReporterFormat: format,
		ListOnly:       listOnly,
		DryRun:         dryRun,
	}, nil
}

func nextArg(args []string, index *int) string {
	*index = *index + 1
	if *index >= len(args) {
		return ""
	}
	return args[*index]
}

func peekArg(args []string, index int) string {
	if index < 0 || index >= len(args) {
		return ""
	}
	return args[index]
}

func expectFlagValue(flag string, value string) (string, error) {
	if value == "" || strings.HasPrefix(value, "-") {
		return "", fmt.Errorf("%s expects a value", flag)
	}
	return value, nil
}

func parseReporterFormat(value string) (TestReporterFormat, error) {
	switch value {
	case "doc":
		return reporterDoc, nil
	case "progress":
		return reporterProgress, nil
	case "tap":
		return reporterTap, nil
	case "json":
		return reporterJSON, nil
	default:
		return "", fmt.Errorf("unknown --format value '%s' (expected doc, progress, tap, or json)", value)
	}
}

func parsePositiveInt(value string, flag string, min int) (int, error) {
	parsed, err := strconv.Atoi(value)
	if err != nil || parsed < min {
		return 0, fmt.Errorf("%s expects an integer >= %d", flag, min)
	}
	return parsed, nil
}

func generateShuffleSeed() int64 {
	now := time.Now().UnixMilli()
	str := fmt.Sprintf("%d", now)
	if len(str) > 9 {
		str = str[len(str)-9:]
	}
	parsed, _ := strconv.ParseInt(str, 10, 64)
	return parsed
}

func resolveTestTargets(targets []string) ([]string, error) {
	rawTargets := targets
	if len(rawTargets) == 0 {
		rawTargets = []string{"."}
	}
	seen := make(map[string]struct{})
	var resolved []string
	for _, target := range rawTargets {
		abs, err := filepath.Abs(target)
		if err != nil {
			return nil, fmt.Errorf("unable to resolve %s: %w", target, err)
		}
		info, err := os.Stat(abs)
		if err != nil {
			return nil, fmt.Errorf("unable to access %s: %w", abs, err)
		}
		if info.IsDir() {
			if _, ok := seen[abs]; !ok {
				seen[abs] = struct{}{}
				resolved = append(resolved, abs)
			}
			continue
		}
		if info.Mode().IsRegular() {
			if isTestFile(abs) {
				if _, ok := seen[abs]; !ok {
					seen[abs] = struct{}{}
					resolved = append(resolved, abs)
				}
			} else {
				dir := filepath.Dir(abs)
				if _, ok := seen[dir]; !ok {
					seen[dir] = struct{}{}
					resolved = append(resolved, dir)
				}
			}
			continue
		}
		return nil, fmt.Errorf("unsupported test target: %s", abs)
	}
	return resolved, nil
}

func collectTestFiles(targets []string) ([]string, error) {
	found := make(map[string]struct{})
	for _, target := range targets {
		info, err := os.Stat(target)
		if err != nil {
			return nil, err
		}
		if info.IsDir() {
			if err := walkTestFiles(target, found); err != nil {
				return nil, err
			}
			continue
		}
		if isTestFile(target) {
			found[filepath.Clean(target)] = struct{}{}
		}
	}
	files := make([]string, 0, len(found))
	for file := range found {
		files = append(files, file)
	}
	sort.Strings(files)
	return files, nil
}

func walkTestFiles(root string, found map[string]struct{}) error {
	return filepath.WalkDir(root, func(path string, d fs.DirEntry, err error) error {
		if err != nil {
			return err
		}
		if d.IsDir() {
			switch d.Name() {
			case "quarantine", "node_modules", ".git":
				return fs.SkipDir
			default:
				return nil
			}
		}
		if d.Type().IsRegular() && isTestFile(d.Name()) {
			found[filepath.Clean(path)] = struct{}{}
		}
		return nil
	})
}

func isTestFile(path string) bool {
	return strings.HasSuffix(path, ".test.able") || strings.HasSuffix(path, ".spec.able")
}
